src/Microsoft.ML.StandardLearners/Standard/MultiClass/MetaMulticlassTrainer.cs

```diff
@@ -21,13 +22,12 @@ public abstract class MetaMulticlassTrainer<TPred, TArgs> : TrainerBase<TPred>
     {
         public abstract class ArgumentsBase
         {
-            [Argument(ArgumentType.Multiple, HelpText = "Base predictor", ShortName = "p", SortOrder = 1)]
+            [Argument(ArgumentType.Multiple, HelpText = "Base predictor", ShortName = "p", SortOrder = 1, SignatureType = typeof(SignatureBinaryClassifierTrainer))]
             [TGUI(Label = "Predictor Type", Description = "Type of underlying binary predictor")]
-            public SubComponent<TScalarTrainer, SignatureBinaryClassifierTrainer> PredictorType =
-                new SubComponent<TScalarTrainer, SignatureBinaryClassifierTrainer>(LinearSvm.LoadNameValue);
+            public IComponentFactory<TScalarTrainer> PredictorType;
```

Comment:
>PredictorType [](start = 53, length = 13)

This is force us to go and put ITrainer<IPredictorProducing<Float>> to all our binary learners.
For example FastTree is just ITrainer in current moment.
Which I think is a good thing, but we need to do it, overwise we can't use OVA in same way as before this change.

Reply:
Actually, I agree it's a good thing that we need to make our trainers produce IPredictorProducing

---
In reply to: [208768180](https://github.com/dotnet/machinelearning/pull/622#discussion_r208768180) [](ancestors = 208768180)

Reply:
Did an internal test catch this?  Or did you find it some other way?

Reply:
All internal tests use old way of instantiating data, via string :)
I've tried to write OVA example with new changes, and found I can't pass anything here, because nothing satisfy to this interface.


---
In reply to: [208774082](https://github.com/dotnet/machinelearning/pull/622#discussion_r208774082) [](ancestors = 208774082)

Reply:
That was the idea with the `SimpleComponentFactory` class(es):

```C#
            using (var env = new TlcEnvironment())
            {
                var ovaArgs = new Ova.Arguments();
                ovaArgs.PredictorType = new SimpleComponentFactory<ITrainer<IPredictorProducing<float>>>(
                    (e) => new FastTreeBinaryClassificationTrainer(e, new FastTreeBinaryClassificationTrainer.Arguments()));

                var ova = new Ova(env, ovaArgs);
            }
```

Reply:
can you put this code into setter for PredictorType?
public IComponentFactory<TScalarTrainer> PredictorType = new SimpleComponentFactory<ITrainer<IPredictorProducing<float>>>((e)=> new LinearSvm(e, new LinearSvm.Arguments())));

So we will have example of how it's done.

Also I personally prefer to throw exception if user set PredictorType to null, rather than falling to default value.


---
In reply to: [209000280](https://github.com/dotnet/machinelearning/pull/622#discussion_r209000280) [](ancestors = 209000280)

Reply:
The reason I did it the way I did was because it skips using the Factory all together in the default case. No need to create an extra factory object or a delegate - if the user didn't provide a PredictorType, we just use `LinearSvm`.

It also solves the case where the user set the field to `null`, as an added bonus. There's no reason to throw an exception in this case. We have a default we can use if the user doesn't give us one.